libnavigation-core/src/test/java/com/mapbox/navigation/core/CurrentIndicesSnapshotProviderTest.kt

```diff
@@ -0,0 +1,224 @@
+            every { currentRouteGeometryIndex } returns routeGeometryIndex2
+            every { currentLegProgress } returns mockk {
+                every { legIndex } returns legIndex2
+                every { geometryIndex } returns legGeometryIndex2
+            }
+        }
+        val expected = CurrentIndicesSnapshot(legIndex1, routeGeometryIndex1, legGeometryIndex1)
+        provider.onRouteProgressChanged(routeProgress1)
+
+        launch {
+            delay(500)
+            provider.onRouteProgressChanged(routeProgress2)
+        }
+
+        assertEquals(expected, provider.getFilledIndicesAndFreeze())
+        assertEquals(expected, provider())
+    }
+
+    @Test
+    fun waitsForUpdateIfValueIsCleared() = runBlocking {
+        val legIndex1 = 5
+        val routeGeometryIndex1 = 78
+        val legGeometryIndex1 = 61
+        val routeProgress1 = mockk<RouteProgress> {
+            every { currentRouteGeometryIndex } returns routeGeometryIndex1
+            every { currentLegProgress } returns mockk {
+                every { legIndex } returns legIndex1
+                every { geometryIndex } returns legGeometryIndex1
+            }
+        }
+        val legIndex2 = 9
+        val routeGeometryIndex2 = 44
+        val legGeometryIndex2 = 33
+        val routeProgress2 = mockk<RouteProgress> {
+            every { currentRouteGeometryIndex } returns routeGeometryIndex2
+            every { currentLegProgress } returns mockk {
+                every { legIndex } returns legIndex2
+                every { geometryIndex } returns legGeometryIndex2
+            }
+        }
+        val expected = CurrentIndicesSnapshot(legIndex2, routeGeometryIndex2, legGeometryIndex2)
+        provider.onRouteProgressChanged(routeProgress1)
+
+        provider.clear()
+        launch {
+            delay(100)
+            provider.onRouteProgressChanged(routeProgress2)
+        }
+
+        assertEquals(expected, provider.getFilledIndicesAndFreeze())
+        assertEquals(expected, provider())
+    }
+
+    @Test
+    fun getFilledIndicesAndFreezeForbidsUpdates() = runBlocking {
+        provider.onRouteProgressChanged(routeProgress)
+        val expected = provider.getFilledIndicesAndFreeze()
+
+        provider.onRouteProgressChanged(mockk {
+            every { currentRouteGeometryIndex } returns routeGeometryIndex
+            every { currentLegProgress } returns mockk {
+                every { legIndex } returns 89
+                every { geometryIndex } returns 12
+            }
+        })
+
+        assertEquals(expected, provider())
+        assertEquals(expected, provider.getFilledIndicesAndFreeze())
+    }
+
+    @Test
+    fun invokeDoesNotForbidUpdates() = runBlocking {
+        provider()
+        provider.onRouteProgressChanged(routeProgress)
+
+        assertEquals(expected, provider())
+        assertEquals(expected, provider.getFilledIndicesAndFreeze())
+    }
+
+    @Test
+    fun unfreezeAllowsUpdates() = runBlocking {
```

Comment:
This also should have an async, cancellation-based test variant. However, since we don't consider this a true use-case (I don't think it would work very well), we can skip it in favor of https://github.com/mapbox/mapbox-navigation-android/issues/6182 work.